lib/docker2aci.go

```diff
@@ -610,112 +621,96 @@ func writeManifest(outputWriter *tar.Writer, manifest schema.ImageManifest) erro
+	if parsedDockerURL.Tag != "" {
+		squashedFilename += "-" + parsedDockerURL.Tag
 	}
+	squashedFilename += ".aci"
 
-	return manifests, nil
+	return squashedFilename
 }
 
-func getFilesToLayersMap(layers []string) (map[string]string, error) {
-	var err error
-	fileMap := make(map[string]string)
-	for _, aciPath := range layers {
-		fileMap, err = gatherFilesToLayersMap(fileMap, aciPath)
+func getManifests(renderedACI acirenderer.RenderedACI, aciRegistry acirenderer.ACIRegistry) ([]schema.ImageManifest, error) {
+	var manifests []schema.ImageManifest
+
+	for _, aci := range renderedACI {
+		im, err := aciRegistry.GetImageManifest(aci.Key)
 		if err != nil {
 			return nil, err
 		}
+		manifests = append(manifests, *im)
 	}
 
-	return fileMap, nil
-}
-
-// gatherFilesToLayersMap accumulates a map associationg each file of the final
-// image with the layer it comes from. It should be called starting from the
-// base layer so that the order of files in the squashed layer is preserved
-// and, if a file is present several times, the last layer is taken into
-// account.
-func gatherFilesToLayersMap(fileMap map[string]string, currentPath string) (map[string]string, error) {
-	currentFile, err := os.Open(currentPath)
-	if err != nil {
-		return nil, err
-	}
-	defer currentFile.Close()
-
-	reader, err := aci.NewCompressedTarReader(currentFile)
-	if err != nil {
-		return nil, err
-	}
-
-	if err = tarball.Walk(*reader, func(t *tarball.TarFile) error {
-		if t.Name() == "manifest" {
-			return nil
-		}
-
-		fileMap[t.Name()] = currentPath
-		return nil
-	}); err != nil {
-		return nil, err
-	}
-
-	return fileMap, nil
+	return manifests, nil
 }
 
-func writeSquashedImage(outputFile *os.File, layers []string, fileMap map[string]string, manifests []schema.ImageManifest) error {
+func writeSquashedImage(outputFile *os.File, renderedACI acirenderer.RenderedACI, aciProvider acirenderer.ACIProvider, manifests []schema.ImageManifest) error {
 	outputWriter := tar.NewWriter(outputFile)
 	defer outputWriter.Close()
 
-	var err error
-	for _, aciPath := range layers {
-		outputWriter, err = reduceACIs(outputWriter, fileMap, aciPath)
+	for _, aciFile := range renderedACI {
+		rs, err := aciProvider.ReadStream(aciFile.Key)
 		if err != nil {
 			return err
 		}
+		defer rs.Close()
+
+		tr := tar.NewReader(rs)
+
+		if err = tarball.Walk(*tr, func(t *tarball.TarFile) error {
+			cleanName := filepath.Clean(t.Name())
+
+			if _, ok := aciFile.FileMap[cleanName]; ok {
+				// we generate and add the squashed manifest later
+				if cleanName == "manifest" {
+					return nil
+				}
+				if err := outputWriter.WriteHeader(t.Header); err != nil {
+					return fmt.Errorf("error writing header: %v", err)
+				}
+				if _, err := io.Copy(outputWriter, t.TarStream); err != nil {
+					return fmt.Errorf("error copying file into the tar out: %v", err)
+				}
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
```

Comment:
not a huge fan of this error handling style, could you pull the walker out into a named function?